### PR TITLE
Substitute drv outputs

### PR DESCRIPTION
**Pull request by regnat**
~~Depends on #4225, only commits after 97aca21 are new.~~

Depends on #4487 

Changes the build loop to allow substituting derivation outputs.

### CODE REVIEW

File: src/libstore/daemon.cc [hunk 1]
```diff
@@ -857,6 +857,27 @@ static void performOp(TunnelLogger * logger, ref<Store> store,
         break;
     }
 
+    case wopRegisterDrvOutput: {
+        logger->startWork();
+        auto outputId = DrvOutputId::parse(readString(from));
+        auto outputPath = StorePath(readString(from));
+        store->registerDrvOutput(outputId, DrvOutputInfo{outputPath});
```
[the remark this answers is not in the document]
Comment (regnat) [reply src/libstore/daemon.cc:864]: It definitely should, yes, and that's one of the necessary follow-ups. 

I think it's okay to keep it for a subsequent PR though, as long as it is behind the `ca-derivations` feature flag (which it isn't atm, I'll have to add it)

[the remark this answers is not in the document]
Comment (roberth) [reply src/libstore/daemon.cc:864]: Where do you track these problems?

[the remark this answers is not in the document]
Comment (regnat) [reply src/libstore/daemon.cc:864]: Curently mostly on a sheet of paper on my desk. Opening proper issues is still on my todo-list

[the remark this answers is not in the document]
Comment (regnat) [reply src/libstore/daemon.cc:864]: I've fixed it in 3be2780 (as part of #4225). Now `registerDrvOutput` requires the experimental feature `ca-derivations` to be enabled if we can't prove that it's indeed a valid output of an input-addressed derivation.

I've also opened #4248 to track the signature problem

**Comment by Ericson2314** [comment]
Getting into the weeds, I think some of the `inputsRealised` logic on resolving would need to be factored out so it can also be used twice`.

Basically we end up with:

1. Try to fetch the outputs.
2. Try to resolve the inputs. If we can, then resolve the derivation, and delegate to building that (call bits factored out from `inputsRealised` call)
3. Try to download the inputs (same)
4. Resolve the derivation (`inputsRealised`, and same logic, but now calls factored-out bits to do it)

This can be done in a follow-up PR, of course.

**Comment by regnat** [comment]
> @regnat I don't think this depends on anything anymore?

It still depends #4487 which itself depends on #4340 and #4372 (unless we want to merge it **into** #4487. I don't really care about the order as long as everything ends-up merged)

> Also, there is another place we want to substitute resolutions: when resolving CA derivations. In that case we _don't_ want to fetch the data. That means we would need to not have resolution goals spawn path goals, at least not unconditionally.
> Getting into the weeds, I think some of the inputsRealised logic on resolving would need to be factored out so it can also be used twice`.

Fully agree. I've opened #3393 to track that (sorry, I though there was already an issue about it)

**Pull request closed without merging**